proposals/async-method-builders.md

```diff
@@ -46,23 +46,69 @@ namespace System.Runtime.CompilerServices
     }
 }
 ```
 
-In the C# compiler, prefer the attribute on the method when determining what builder to use over the one defined on the type.  For example, today if a method is defined as:
+The attribute can be applied on methods, types and modules.
+
+Example of usage on a method:  
+```C#
+[AsyncMethodBuilderOverride(typeof(PoolingAsyncValueTaskMethodBuilder<int>))] // new, referring to some custom builder type
+static async ValueTask<int> ExampleAsync() { ... }
+```
+
+A developer that wants to using a specific custom builder for all of their methods can do so by putting the relevant attribute on each method.  
+Example of usage on module:  
+```C#
+[module: AsyncMethodBuilderOverride(typeof(PoolingAsyncValueTaskMethodBuilder), typeof(ValueTask)]
+[module: AsyncMethodBuilderOverride(typeof(PoolingAsyncValueTaskMethodBuilder<>), typeof(ValueTask<>)]
+
+class MyClass
+{
+    public async ValueTask Method1Async() { ... } // would use PoolingAsyncValueTaskMethodBuilder
+    public async ValueTask<int> Method2Async() { ... } // would use PoolingAsyncValueTaskMethodBuilder<int>
+    public async ValueTask<string> Method3Async() { ... } // would use PoolingAsyncValueTaskMethodBuilder<string>
+}
+```
+
+### Determining the builder type for an async methd
+
+When compiling an async method, the builder type is determined by:
+1. looking in the containing scopes for an override attribute that specifies a builder type compatible with the method's return type.
+2. otherwise, falling back to the builder type determined by previous approach. (see [spec for task-like types](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.0/task-types.md)) 
+
+The containing scope are considered starting from the method, then its containing types and finally the module. Each scope is considered
+in turn.  
+If a scope has one or more override attributes, the following process is used to determine if it is compatible with the current async method:
+1. take the override type specified by the attribute and construct it if necessary.  
+  If the override type is an open generic type, take the single type argument of the async method's return type and substitute it into the override type. If
+  the async method's return type does not have a single type argument, then that override is not compatible.
+2. look for the public `Create` method with no type parameters and no parameters on the constructed override type:  
+  If the method is not found, then that override is not compatible.
+3. consider the return type of that `Create` method (a builder type) and look for the public `Task` property.  
+  If the property is not found, then that override is not compatible.
+4. consider the type of that `Task` property (a task-like type):  
+  If the task-like type matches the return type of the async method, then the override is compatible. Otherwise, it is not compatible.
+
+If no compatible override is found on a given scope, then look at the next scope.  
```

Comment:
>If no compatible override is found on a given scope, then look at the next scope. [](start = 0, length = 81)

If the attribute is applied to a method directly but is incompatible, should we report an error instead of silently ignoring that attribute?

Reply:
Added special rule when the attribute is applied to method. Thanks